Mathlib/Algebra/PresentedMonoid/Basic.lean

```diff
@@ -0,0 +1,248 @@
+/-
+Copyright (c) 2024. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hannah Fechtner
+-/
+
+import Mathlib.Algebra.FreeMonoid.Basic
+import Mathlib.Algebra.Group.Submonoid.Operations
+
+/-!
+# Defining a monoid given by generators and relations
+
+Given a subset `rels` of relations of the free monoid on a type `α`, this file constructs the monoid
+given by generators `x : α` and relations `r ∈ rels`.
+
+## Main definitions
+
+* `PresentedMonoid rels`: the quot group of the free group on a type `α` by the steps-to closure
+  of a subset `rels` of relations of the free monoid on `α`.
+* `of`: The canonical map from `α` to a presented monoid with generators `α`.
+* `toMonoid f`: the canonical monoid homomorphism `PresentedMonoid rels → M`, given a function
+  `f : α → G` from a type `α` to a monoid `M` which satisfies the relations `rels`.
+
+## Tags
+
+generators, relations, monoid presentations
+-/
+
+variable {α : Type*}
+
+/-- the steps-to closure of a relation -/
+inductive PresentedMonoid.StepsTo {β : Type*} [Mul β] (R : β → β → Prop) : β → β → Prop
+  | basic : ∀ {x y}, R x y → StepsTo R x y
+  | left  : ∀ {x y z}, StepsTo R x y → StepsTo R (z * x) (z * y)
+  | right : ∀ {x y z}, StepsTo R x y → StepsTo R (x * z) (y * z)
+
+/-- Given a set of relations, `rels`, over a type `α`, `PresentedMonoid` constructs the monoid with
+generators `x : α` and relations `rels` as a quot of `FreeMonoid α`. -/
+def PresentedMonoid (rel : FreeMonoid α → FreeMonoid α → Prop) :=
+  Quot (PresentedMonoid.StepsTo rel)
```

Comment:
I still think `abbrev PresentedMonoid := (conGen rel).Quotient` is a better definition. It gives you the monoid structure and the surjection `FreeMonoid α -> PresentedMonoid rel` and the universal property for free.

Reply:
done!